### PR TITLE
Working bdd

### PR DESCRIPTION
**Pull request by whereismyai**

### CODE REVIEW

File: src/main/java/pages/CataloguePage.java [hunk 1]
```diff
@@ -1,59 +1,178 @@
...
+import org.openqa.selenium.By;
 import org.openqa.selenium.NoSuchElementException;
 import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
 import org.openqa.selenium.support.FindBy;
 
 import java.util.List;
 
-public class CataloguePage extends BasePage {
+public class CataloguePage extends HeaderAnyPage {
```
Comment (DmitriyBondarenko) [inline src/main/java/pages/CataloguePage.java]: Why do we need to extend `HeaderAnyPage` class here? BTW please, rename `HeaderAnyPage` to `BaseHeader`

File: src/main/java/pages/ComparingPage.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 import java.util.List;
 
-public class ComparingPage extends BasePage {
+public class ComparingPage extends HeaderAnyPage {
```
Comment (DmitriyBondarenko) [inline src/main/java/pages/ComparingPage.java]: Same

File: src/main/java/pages/ComparingPage.java [hunk 3]
```diff
@@ -41,20 +41,13 @@ public static String getComparingLink() {
         return comparingLink;
     }
 
-    public int amountOfComparingProducts() {
+    public String amountOfComparingProducts() {
         wait.tillElementsPresent(productsOnComparing);
-        return productsOnComparing.size();
+        return String.valueOf(productsOnComparing.size());
```
Comment (DmitriyBondarenko) [inline src/main/java/pages/ComparingPage.java]: Method should return 'amount', but you want to return string? Why do we need this?

File: src/main/java/pages/MainPage.java [hunk 4]
```diff
@@ -1,22 +1,28 @@
...
 import org.openqa.selenium.WebElement;
 import org.openqa.selenium.support.FindBy;
 import util.dataUtils.CharDataForTestSite;
 
 import java.util.List;
 
-public class MainPage extends BasePage {
+import static java.lang.String.format;
+
+public class MainPage extends HeaderAnyPage {
```
Comment (DmitriyBondarenko) [inline src/main/java/pages/MainPage.java]: Same

File: src/main/java/steps/CatalogueSteps.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    public CatalogueSteps(DriverProvider driverProvider) {
+        this.driverProvider = driverProvider;
+        wait = new WaitUtils(driverProvider.getDriver());
+    }
+
+    @Then("Catalogue is displayed")
+    public void catalogueIsDisplayed() {
+        cataloguePage = new CataloguePage(driverProvider.getDriver());
+        assertThat(
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: Code format

File: src/main/java/steps/CatalogueSteps.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    private DriverProvider driverProvider;
+    private List<ProductDto> products = new ArrayList<>();
+
+
+    public CatalogueSteps(DriverProvider driverProvider) {
+        this.driverProvider = driverProvider;
+        wait = new WaitUtils(driverProvider.getDriver());
+    }
+
+    @Then("Catalogue is displayed")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: `Catalogue page is displayed`

File: src/main/java/steps/CatalogueSteps.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        cataloguePage.LOWPriceFilterField.sendKeys(String.valueOf(low));
+        cataloguePage.HIGHPriceFilterField.sendKeys(String.valueOf(high));
+        cataloguePage.OKButtonPriceFilter.click();
+        wait.tillElementInvisible(cataloguePage.catalogue.findElement(By.xpath(".//div[contains(@class, 'on-loading')]")));
+        updateProductList();
+    }
+
+    @Then("Only products with a price in range from {int} to {int} are shown")
+    public void onlyProductsWithAPriceInRangeFromLowToHighAreShown(int low, int high) {
+        assertThat(
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: Code format

File: src/main/java/steps/CatalogueSteps.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    public void userFilterProductsByPriceInRangeFromLowToHigh(int low, int high) {
+        cataloguePage = new CataloguePage(driverProvider.getDriver());
+        cataloguePage.LOWPriceFilterField.sendKeys(String.valueOf(low));
+        cataloguePage.HIGHPriceFilterField.sendKeys(String.valueOf(high));
+        cataloguePage.OKButtonPriceFilter.click();
+        wait.tillElementInvisible(cataloguePage.catalogue.findElement(By.xpath(".//div[contains(@class, 'on-loading')]")));
+        updateProductList();
+    }
+
+    @Then("Only products with a price in range from {int} to {int} are shown")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: `Products with a price range from {int} to {int} are shown`

File: src/main/java/steps/CatalogueSteps.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @When("User using filter to see the products with price more than {int}")
+    public void userUsingFilterToSeeTheProductsWithPriceMoreThanLow(int low) {
+        cataloguePage.LOWPriceFilterField.sendKeys(String.valueOf(low));
+        cataloguePage.OKButtonPriceFilter.click();
+        updateProductList();
+    }
+
+    @Then("Only products with a price  more than {int} are shown")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: `Products with a price more than {int} are shown`

File: src/main/java/pages/CataloguePage.java (hunk 1 above)
Comment (E-Kel) [reply src/main/java/pages/CataloguePage.java]: ok. to make possible to work with header (serarch and user account actions) from any page.

File: src/main/java/steps/CatalogueSteps.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    @Then("Only products with a price in range from {int} to {int} are shown")
+    public void onlyProductsWithAPriceInRangeFromLowToHighAreShown(int low, int high) {
+        assertThat(
+                products.isEmpty()
+                        || ProductDto.getLowestPrice(products) >= low
+                        && ProductDto.getHighestPrice(products) <= high)
+                .isTrue();
+    }
+
+    @When("User using filter to see the products with price more than {int}")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: `User applies filter with price more than {int}` or smth like that

File: src/main/java/steps/CatalogueSteps.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        updateProductList();
+    }
+
+    @Then("Only products with a price  more than {int} are shown")
+    public void onlyProductsWithAPriceMoreThanLowAreShown(int low) {
+        assertThat(products.stream().allMatch(productDto -> (productDto.getLowestPrice() >= low)));
+
+    }
+
+    @When("User using filter to see the products with price less than {int}")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/CatalogueSteps.java]: Same as above

File: src/main/java/steps/ComparingSteps.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        wait = new WaitUtils(driverProvider.getDriver());
+        mainPage = new MainPage(driverProvider.getDriver());
+    }
+
+    @And("User add {string} products to comparing")
+    public void userAddProductsToComparing(String amount) {
+        cataloguePage.addProductsToComparing(Integer.parseInt(amount));
+    }
+
+    @And("Clicks Compare button")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/ComparingSteps.java]: `User clicks Compare button`

File: src/main/java/steps/NavigationSteps.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+public class NavigationSteps {
+    protected DriverProvider driverProvider;
+    protected MainPage mainPage;
+    protected CataloguePage cataloguePage;
+    private WaitUtils wait;
+
+
+    public NavigationSteps(DriverProvider driverProvider) {
+
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/NavigationSteps.java]: Remove LB

File: src/main/java/steps/NavigationSteps.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import pages.MainPage;
+import util.driverUtils.DriverProvider;
+import util.elementUtils.WaitUtils;
+
+public class NavigationSteps {
+    protected DriverProvider driverProvider;
+    protected MainPage mainPage;
+    protected CataloguePage cataloguePage;
+    private WaitUtils wait;
+
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/NavigationSteps.java:10]: Remove LB

File: src/main/java/steps/SearchSteps.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    private WaitUtils wait;
+
+
+    public SearchSteps(DriverProvider driverProvider) {
+        wait = new WaitUtils(driverProvider.getDriver());
+        mainPage = new MainPage(driverProvider.getDriver());
+        this.driverProvider = driverProvider;
+    }
+
+    @When("Enter {string} into search field and press enter")
```
Comment (DmitriyBondarenko) [inline src/main/java/steps/SearchSteps.java]: Maybe `User searches for {string}`? Or smth like that

File: src/main/java/util/driverUtils/DriverProvider.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+    protected WebDriver driver;
+
+    public WebDriver getDriver() {
+        return driver;
+    }
+
+    public void setupController() {
+        this.driver = BrowserFactory.getDriver();
+        this.driver.manage().window().maximize();
```
Comment (DmitriyBondarenko) [inline src/main/java/util/driverUtils/DriverProvider.java]: We already have `start-maximized` parameter in `getDriver()` method

**Review by DmitriyBondarenko — CHANGES_REQUESTED** [review]
Comments

**Pull request merged**